Add explicit overload to `Field` subclasses

Viicos opened the pull request:
In https://github.com/typeddjango/django-stubs/issues/1264#issuecomment-1859762279, and to try working on bringing `django-stubs` and `django-types` together, I suggested a workaround that would allow type checkers to natively infer `_ST` and `_GT` without the need for a plugin.

Unfortunately, for it to work correctly it would most probably require [higher kinded type vars](https://github.com/python/typing/issues/548) (my current example doesn't play well with subclasses, and I can't use `Self` as it can't be parametrized).

This is an experimental test, and as you can see it adds a lot of verbosity. But this removes the need to a plugin to parametrize the type variables, and it support nullable fields.

Would you accept such a change? I'm ok if this isn't accepted, considering how much it adds

sobolevn reviewed (commented):
I like the idea. Several comments:
1. Please, delete the plugin code, so we can be sure that your changes are enough to pass our CI
2. Looks like `stubstest` is not happy with some of the changes. Please, fix or ignore them :)

Viicos commented on django-stubs/contrib/auth/models.pyi:
```diff
@@ -23,9 +24,9 @@ class Permission(models.Model):
     content_type_id: int
     objects: ClassVar[PermissionManager]
 
-    name = models.CharField(max_length=255)
-    content_type = models.ForeignKey(ContentType, on_delete=models.CASCADE)
-    codename = models.CharField(max_length=100)
```
Having explicit assignments feels a bit weird in a stub file and was causing issues when testing without the plugin (had error code `var-annotated`)

Viicos commented on django-stubs/contrib/auth/models.pyi:
```diff
@@ -34,7 +35,7 @@ class GroupManager(models.Manager[Group]):
 
-    name = models.CharField(max_length=150)
+    name: models.CharField[str | int | Combinable, str]
     permissions = models.ManyToManyField(Permission)
```
... however I can't explicitly annotate `permissions: models.ManyToManyField[..., ...]` as the second type var is solved by the plugin to a class that doesn't exist in stubs nor at runtime

flaeppe replied:
You can have a look here how it can be done: https://github.com/typeddjango/django-stubs/pull/2214/files#diff-0236ca9e909bc4743fa03e6653fc33adbaaef02a34d8698501a63a59ad624f1fR42-R51

Viicos commented on tests/typecheck/fields/test_base.yml:
```diff
@@ -1,4 +1,15 @@
+                    mypy_django_plugin.main
+        -   mypy_section: |
+                [mypy]
+                disable_error_code = var-annotated
```
...So I had to disable the error code anyway when testing without the plugin

Viicos committed:
Add explicit overloads to each field class

This allows type checkers other than mypy to infer correctly
the `__set/get__` type variables, without the need for the plugin.

The plugin logic to solve these type variables is kept, as some
special casing is implemented to handle primary key fields with a default.
The `Field` docstring was updated with an example reflecting this

Viicos committed:
Explicitly annotate some fields in the auth module

This is required to run mypy without the plugin, as it would complain
about these fields requiring an annotation. Using `follow_imports = silent`
did not work in this case

Also feels more natural than doing an assignment in a stub file

Viicos committed:
Add tests to ensure this is working without the plugin

A bit of a hacky solution for now, maybe CI could be improved to
add this kind of tests more easily.
`var-annotated` is still an ignored error code as I wasn't able
to change the assignment to an annotation for `ManyToManyField`s
in my previous commit

Viicos committed:
Add fields __init__ to allowlist

Really not ideal, so open to alternatives

Viicos committed:
Incorporate latest changes

`db_default` is added on all overloads. `_ST` is replaced
by the same set type set on `self`

Viicos committed:
Fix newly added tests

Viicos committed:
wip: Do not infer types from `_pyi_private_*_type`

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
